Add star into link attributes

tlakollo opened the pull request:
One of the functionalities of the LinkAttributes files is to occurrences of custom attributes on the current assembly, on top of this functionality there is a special functionality that is only permitted if it's embedded in an XML inside the System.Private.CorLib module that allows the deletion of custom attributes from all the assemblies.
Given that NativeAOT is multithreaded it cannot process all the assemblies to remove attributes as ILLink does, it also does not have a shared storage in which it can store the information about the removed attributes from System.Private.CoreLib.
This PR adds logic to process the special functionality in the embedded System.Private.CoreLib XML every time an assembly is being processed. 
In order to make this more performant, an extra argument is passed to the XML reader to only look at the assembly="*" to get the set of attributes that need to be deleted from all assemblies.

Fixes https://github.com/dotnet/runtime/issues/77753

Commit:
Initial commit of handling the star argument in LinkAttributes files for NativeAOT.
Add testing infrastructure from illink

Commit:
Remove linker testing since it tests more unsupported scenarios and no actual testing of the AllAssembly functionality
Refactor some of the code

Commit:
Add a way to only process star/nonstar assemblies in ProcessLinkerXmlBase
Add test

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -39,7 +39,7 @@ public sealed class UsageBasedMetadataManager : GeneratingMetadataManager
         internal readonly UsageBasedMetadataGenerationOptions _generationOptions;
 
-        private readonly FeatureSwitchHashtable _featureSwitchHashtable;
+        private readonly LinkAttributesHashTable _featureSwitchHashtable;
```
Let's get rid of all of the bad copypasting

```suggestion
        private readonly LinkAttributesHashTable _linkAttributesHashTable;
```

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -1126,19 +1126,30 @@ public AssemblyFeatureInfo(EcmaModule module, Logger logger, IReadOnlyDictionary
+
+            public void  ParseLinkAttributesXml(EcmaModule module, Logger logger, IReadOnlyDictionary<string, bool> featureSwitchValues, bool globalAttributeRemoval)
+            {
+                EcmaModule xmlModule = globalAttributeRemoval ? (EcmaModule) module.Context.SystemModule : module;
```
I rarely have to comment on formatting but the IL Linker formatting rules appearing where they shouldn't is starting to be a nuisance (this comment is not directly on your PR - I've commented on these in other PRs of people who work in both codebases in the past weeks, cc @vitek-karas).

There's an undeniable overhead in switching between styles. This is starting to be a friction point in PRs. Formatting never was a friction point. Honestly the linker coding style is creating a problem that I haven't seen in my 10 years on the team.

```suggestion
                EcmaModule xmlModule = globalAttributeRemoval ? (EcmaModule)module.Context.SystemModule : module;
```

vitek-karas replied:
We should switch to linker in runtime repo real soon (hopefully this week). Once that is done I think the first thing we do on the linker is to update the formatting to match runtime rules. And then all of this will be history...

MichalStrehovsky replied:
> the first thing we do on the linker is to update the formatting to match runtime rules

Before doing that, it might be good to catch up parts of NativeAOT that are line-by-line ports of ILLink and delete ReferenceSource. Reformatting will introduce lots of diffs to ReferenceSource. It's not impossible to do the catchup later, but it will be an extra annoyance. Once it's all caught up, we don't need ReferenceSource anymore - we just have to be diligent in PRs to update both files always.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -1150,37 +1161,68 @@ public AssemblyFeatureInfo(EcmaModule module, Logger logger, IReadOnlyDictionary
+        internal sealed class LinkAttributesReader : ProcessLinkerXmlBase
         {
             private readonly HashSet<TypeDesc> _removedAttributes = new();
+            private readonly ModuleDesc _resource;
```
I think this is available as `_owningModule`

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -1150,37 +1161,68 @@ public AssemblyFeatureInfo(EcmaModule module, Logger logger, IReadOnlyDictionary
+                        return AllowedAssemblies.AllAssemblies;
+
+                    // Corelib XML may contain assembly wildcard to support compiler-injected attribute types
+                    if (_resource?.Assembly.GetName().Name == PlatformAssemblies.CoreLib)
```
```suggestion
                    if (_resource?.Assembly == _context.SystemModule)
```

MichalStrehovsky commented on src/tests/nativeaot/SmokeTests/TrimmingBehaviors/ILLinkLinkAttributes.cs:
```diff
@@ -0,0 +1,98 @@
+    
+    class TestDontRemoveAttribute : Attribute
+	{
+		public TestDontRemoveAttribute ()
```
```suggestion
		public TestDontRemoveAttribute()
```

MichalStrehovsky commented on src/tests/nativeaot/SmokeTests/TrimmingBehaviors/ILLinkLinkAttributes.cs:
```diff
@@ -0,0 +1,98 @@
+
+	class TestRemoveAttribute : Attribute
+	{
+		public TestRemoveAttribute ()
```
```suggestion
		public TestRemoveAttribute()
```

Commit:
Run formatter
Apply feedback
Avoid using LogWarning in ReadyToRun

tlakollo committed:
Ask early on if the assembly should be processed, no reason to do work before that
reorg conditions to print warnings only if _globalAttributeRemoval is false
Move assembly to process check for non star members only

MichalStrehovsky reviewed (approved):
Looks great!

tlakollo committed:
Merge branch 'main' into AddStarIntoLinkAttributes

Commit:
Actually fix the merge conflict
